provider/azure/storage.go

```diff
@@ -403,35 +405,63 @@ func (v *azureVolumeSource) AttachVolumes(ctx context.ProviderCallContext, attac
 	return results, nil
 }
 
-const azureDiskDeviceLink = "/dev/disk/azure/scsi1/lun%d"
+const (
+	azureDiskDeviceLink = "/dev/disk/azure/scsi1/lun%d"
+	nvmeDiskDeviceName  = "nvme0n%d" // +2 to actual lun
+)
 
 func (v *azureVolumeSource) attachVolume(
 	vm *armcompute.VirtualMachine,
 	p storage.VolumeAttachmentParams,
 ) (_ *storage.VolumeAttachment, updated bool, _ error) {
 
 	var dataDisks []*armcompute.DataDisk
-	if vm.Properties != nil && vm.Properties.StorageProfile.DataDisks != nil {
-		dataDisks = vm.Properties.StorageProfile.DataDisks
+	var diskControllerType *armcompute.DiskControllerTypes
+	if vm.Properties != nil && vm.Properties.StorageProfile != nil {
+		if vm.Properties.StorageProfile.DataDisks != nil {
+			dataDisks = vm.Properties.StorageProfile.DataDisks
+		}
+		if vm.Properties.StorageProfile.DiskControllerType != nil {
+			diskControllerType = vm.Properties.StorageProfile.DiskControllerType
+		}
+	}
+
+	isNVME := false
```

Comment:
Would prefer that we use a var of type `DiskControllerTypes` and switch on that. The default would be `DiskControllerTypesSCSI`.
We would then also pass that var to addDataDisk() below instead of a bool.

Reply:
sure i'll refactor to that if it makes things clearer, thanks ian!